### PR TITLE
Sg2/samplerstatetype

### PR DESCRIPTION
**Pull request by esmelusina**
Post package split sampler state impl. Basic implementation wrapped, it's a very straightforward type with straightforward type generation.

image

### CODE REVIEW

File: com.unity.sg2/Editor/GraphDeltaRegistry/Generation/Interpreter.cs [hunk 1]
```diff
@@ -363,6 +364,13 @@ private static bool FunctionsAreEqual(ShaderFunction a, ShaderFunction b)
                                 if (tex != null && !defaultTextures.Contains((name, tex)))
                                     defaultTextures.Add((name, tex));
                             }
+
+                            if (port.GetTypeField().GetRegistryKey().Name == SamplerStateType.kRegistryKey.Name)
+                            {
+                                var fieldHandler = port.GetTypeField();
+                                var field = SamplerStateType.UniformPromotion(port.GetTypeField(), container, registry);
+                                inputVariables.Add(field);
+                            }
```
Comment (esmelusina) [inline com.unity.sg2/Editor/GraphDeltaRegistry/Generation/Interpreter.cs:373]: Textures (above) and Samplers are the only cases where an an inline value has to be promoted to a material property. I'll consider abstracting this if we run into situations where that will be helpful, but for now I'm not going to fiddle with this much further.

File: com.unity.sg2/Editor/GraphDeltaRegistry/Registry/SamplerType.cs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+                || field.GetSubField<Aniso>(kAniso) != null
+                || field.GetSubField<bool>(kCompare) != null;
+        }
+
+
+        #endregion
+
+
+        public void BuildType(FieldHandler field, Registry registry)
+        {
```
Comment (esmelusina) [inline com.unity.sg2/Editor/GraphDeltaRegistry/Registry/SamplerType.cs:115]: SamplerStates, in Unity, have a very specific set of behavior requirements. They are parsed as strings and aren't constructed or composed objects. For this reason, we can make a lot of assumptions about SamplerState initialization and avoid any concrete data generation.

File: com.unity.sg2/Editor/GraphDeltaRegistry/Registry/SamplerType.cs [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        public ShaderType GetShaderType(FieldHandler field, ShaderContainer container, Registry registry)
+        {
+            return container._UnitySamplerState;
+        }
+
+        public string GetInitializerList(FieldHandler field, Registry registry)
+        {
+            return $"In.{GetUniqueSamplerName(field)}";
+        }
+
```
Comment (esmelusina) [inline com.unity.sg2/Editor/GraphDeltaRegistry/Registry/SamplerType.cs:127]: Any inline value that is promoted to material property eventually becomes a part of the block input structure, which is always called `In.` So the initial value for a SamplerState reference is going to come from the In struct.

File: com.unity.sg2/Editor/GraphUI/DataModel/Constants/SamplerTypeConstant.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+            SamplerStateType.SetWrap(GetField(), smp.wrap);
+            SamplerStateType.SetDepthComparison(GetField(), smp.depthCompare);
+            SamplerStateType.SetAniso(GetField(), smp.aniso);
+    }
+
+        override public object DefaultValue => Activator.CreateInstance(Type);
+        override public Type Type => typeof(SamplerStateData);
+        override public TypeHandle GetTypeHandle() => ShaderGraphExampleTypes.SamplerStateTypeHandle;
+    }
+
```
Comment (esmelusina) [inline com.unity.sg2/Editor/GraphUI/DataModel/Constants/SamplerTypeConstant.cs:40]: Don't love this- but SamplerStates don't have a public unity standard data object for handling, representing, or manipulating them currently. There's a couple different ways we could setup a constant for them, but this is probably fine and simpler for UI impl.

**Review by brett-groshong — APPROVED** [review]

**Pull request merged**